bitcoinspv/relayer_test.go

```diff
@@ -0,0 +1,192 @@
+		t.Error("Timeout waiting for WaitForShutdown")
+	}
+}
+
+func TestNew(t *testing.T) {
+	relayer, _, _ := setupTest(t)
+
+	assert.NotNil(t, relayer.Config)
+	assert.NotNil(t, relayer.logger)
+	assert.NotNil(t, relayer.btcClient)
+	assert.NotNil(t, relayer.lcClient)
+	assert.Equal(t, time.Second, relayer.Config.RetrySleepDuration)
+	assert.Equal(t, 10*time.Second, relayer.Config.MaxRetrySleepDuration)
+	assert.Equal(t, 5*time.Second, relayer.Config.ProcessBlockTimeout)
+	assert.NotNil(t, relayer.quitChannel)
+	assert.False(t, relayer.isStarted)
+}
+
+func TestIsRunning(t *testing.T) {
+	relayer, _, _ := setupTest(t)
+	assert.False(t, relayer.isRunning())
+
+	relayer.isStarted = true
+	assert.True(t, relayer.isRunning())
+
+	relayer.isStarted = false
+	assert.False(t, relayer.isRunning())
+}
+
+func TestIsShutdown(t *testing.T) {
+	relayer, _, _ := setupTest(t)
+	assert.False(t, relayer.isShutdown())
+
+	// simulate shutdown
+	close(relayer.quitChannel)
+	assert.True(t, relayer.isShutdown())
+
+	// simulate restart
+	relayer.quitChannel = make(chan struct{})
+	assert.False(t, relayer.isShutdown())
+}
+
+func TestStop(t *testing.T) {
+	relayer, _, _ := setupTest(t)
+	initialChan := relayer.quitChan()
+
+	// initial state
+	assert.False(t, relayer.isShutdown())
+	relayer.isStarted = true
+	assert.True(t, relayer.isRunning())
+
+	relayer.Stop()
+
+	// after calling stop
+	assert.True(t, relayer.isShutdown())
+	assert.False(t, relayer.isRunning())
+	_, chanOpen := <-initialChan
+	assert.False(t, chanOpen)
+
+	// call stop again
+	assert.NotPanics(t, func() { relayer.Stop() })
+	assert.True(t, relayer.isShutdown())
+	assert.False(t, relayer.isRunning())
+}
+
+func TestStart(t *testing.T) {
+	relayer, btcClient, lcClient := setupTest(t)
+	setupMocks(t, btcClient, lcClient)
+	t.Cleanup(func() {
+		cleanupRelayer(t, relayer)
+	})
+	assert.False(t, relayer.isRunning())
+
+	go relayer.Start()
+	time.Sleep(10 * time.Millisecond)
+
+	assert.False(t, relayer.isShutdown())
+	assert.True(t, relayer.isRunning())
+
+	// call start again
+	assert.NotPanics(t, func() { relayer.Start() })
+
+	assert.False(t, relayer.isShutdown())
+	assert.True(t, relayer.isRunning())
+
+	btcClient.AssertExpectations(t)
+	lcClient.AssertExpectations(t)
+}
+
+func TestRestartAfterShutdown(t *testing.T) {
+	relayer, btcClient, lcClient := setupTest(t)
+	setupMocks(t, btcClient, lcClient)
+	t.Cleanup(func() {
+		cleanupRelayer(t, relayer)
+	})
+	assert.False(t, relayer.isRunning())
+
+	go relayer.Start()
+	time.Sleep(10 * time.Millisecond)
+
+	assert.True(t, relayer.isRunning())
+	assert.False(t, relayer.isShutdown())
+
+	assert.NotPanics(t, func() { relayer.Stop() })
+	assert.True(t, relayer.isShutdown())
+	assert.False(t, relayer.isRunning())
+
+	// Start again after shutdown
+	go relayer.Start()
+	time.Sleep(10 * time.Millisecond)
+
+	assert.True(t, relayer.isRunning())
+	assert.False(t, relayer.isShutdown())
+
+	btcClient.AssertExpectations(t)
+	lcClient.AssertExpectations(t)
+}
```

Comment:
let's merge this test with the previous one - there is too much overalp

Reply:
Done